Provide tests for `OgRole::isAdmin()` and `OgRole::setIsAdmin()`

pfrenssen opened the pull request:
Fix for #369.

Edit: I was mistaken, the `OgRole::isAdmin()` and `OgRole::setIsAdmin()` methods were not malfunctioning due a bug but due to a misconfiguration in my project.

I've now repurposed this PR to provide some additional test coverage that demonstrates the use of these methods.

pfrenssen committed:
Add test coverage for OgRole::isAdmin() and OgRole::setIsAdmin().

pfrenssen commented:
Travis tests are running but the results don't seem to be reported back.

Here is the failing test: https://travis-ci.org/Gizra/og/builds/380149838
And the fix that turns the test green: https://travis-ci.org/Gizra/og/builds/380152585

amitaibu commented on tests/src/Unit/OgRoleTest.php:
```diff
@@ -0,0 +1,48 @@
+use Drupal\Tests\UnitTestCase;
+
+/**
+ * Tests the OgContext context provider.
```
Wrong comment

amitaibu reviewed (commented):
Looks good, just a tiny comment.

pfrenssen commented:
This apparently breaks something, `OgEntityAccessTest::testAccess()` is failing.

pfrenssen commented:
It's the test that is wrong, instead of setting the role name to `OgRoleInterface::ADMINISTRATOR`, it sets the role name to a random value, and it calls `OgRole::setIsAdmin()` instead.

This points out that we also have to implement `::setIsAdmin()`. We can easily set it (by assigning the value `OgRoleInterface::ADMINISTRATOR` to the role name), but _unsetting_ the admin role by calling `::setIsAdmin(FALSE)` is problematic. We don't know what name to set it to.

I think a good way to handle this case is to set the role name to `OgRoleInterface::AUTHENTICATED` and emitting a PHP warning alongside it, to alert developers that they should instead set the role name directly.

Edit: ~~even better would be to retain the underlying mechanic, and set the value dynamically when the role name is set to `OgRoleInterface::ADMINISTRATOR`.~~ Edit: no, this could put us in the awkward position that a role would be called 'administrator' but is actually not an administrator.

pfrenssen commented:
@amitaibu could it be possible that the Travis CI integration was disabled when the Gizra organisation on Github enabled OAuth app restrictions recently?

amitaibu commented:
I don't think thats the case

![apps_and_integrations_](https://user-images.githubusercontent.com/125707/40182826-f8b1992e-59f4-11e8-9ef6-e717a18574f0.jpg)

@AronNovak have you seen travis not working in other Gizra repos? (I haven't)

pfrenssen commented:
Ah yeah that makes sense, otherwise the tests wouldn't even start I suppose. I just find it strange that the tests results are not reported back.

amitaibu commented:
I suspect it's a travis-ci hickup

AronNovak commented:
@amitaibu :+1: on this, i did not experience any issues at Travis projects. However those are closed projects, I'll check at Elm Starter ASAP.

AronNovak commented:
Hm, actually I used Travis at Elm Starter yesterday, then it must be a hiccup, almost certainly.

pfrenssen commented:
Thanks for having a look @AronNovak !

PR looks good now, tests are passing. What was changed:

* Dropped the storage of the `is_admin` configuration, it wasn't working before, and we don't need it since we can infer it from the role name.
* Implemented `::setIsAdmin()` while throwing a warning to alert developers that this has the unexpected side effect of changing the role name. I also added a test for this implementation.
* Fixed the `OgEntityAccessTest`. The setup of the test was indeed wrong, in reality the admin role is already created when a group is created, so we should use that one.

pfrenssen commented:
Here is the test result of the last commit: https://travis-ci.org/Gizra/og/builds/380237297

amitaibu commented on src/Entity/OgRole.php:
```diff
@@ -269,6 +268,33 @@ public function isRequired() {
+    // @endcode
+    trigger_error('Using ::setIsAdmin() affects the role name.', E_USER_WARNING);
+
+    $this->setName($is_admin ? OgRoleInterface::ADMINISTRATOR : OgRoleInterface::AUTHENTICATED);
```
This doesn't feel right. We say and `admin` is defined by having a role named `admin`, but if it's not an admin, the role name can be whatever we want.

Maybe we should just throw an Exception, if this method is used?

pfrenssen replied:
OK, that's also a good solution, alerting the developer using an exception.

@amitaibu to be 100% clear, it is the intention that a group admin has all permissions in the group, right? Because that is wat `::isAdmin()` promises:

```
  /**
   * Indicates that a role has all available permissions.
   *
   * @return bool
   *   TRUE if the role has all permissions.
   */
  public function isAdmin();
```

Reason I am asking is that in my project we have 1 test failing with this PR applied: we have a use case where the group admin is not allowed to delete content, because the content needs to be retained for archival reasons. With the PR the admin gets this permission.

So my understanding is that this is intentional, and on our end we should implement `hook_og_user_access_alter()` to take away the permission from the group admin?

idimopoulos replied:
ahh, hello :) There is an explicit `is_admin` flag that controls this in core. We can use this. We don't need to restrict it in the administrator role.
The og_roles have inherited this flag and are already in exported configurations. Core also bases this in this flag.

*Edit:*
As I discussed with Pieter in the chat, here are some food for thoughts from our chat.
Currently we are using the `OgRoleInterface::ADMINISTRATOR` constant to define if the user is an administrator. This was the corresponding `user 1` of D7 for groups. My comment was not about the current implementation being wrong. This is merely a design decision which is working fine. My comment was that since core has decided to include an extra flag that makes a role an administrator (and not some global `grant all privileges; flush privileges` permission), by keeping the same administrator named role design, it is still valid, but we are deviating from core. We could use the flag in order to determine the power of the role since it is offered by core and then someone who is accustomed in core and just enters og will find it quite familiar.

pfrenssen replied:
Thanks for the input @idimopoulos! I can see the merit of your proposal. It would simplify this PR, and we would stay closer in spirit to the core implementation, while still retaining the full flexibility.

So to be clear, is this what you are suggesting?

* We keep the `is_admin` flag in exported config. When creating a new group type this flag will be automatically set when the admin role is auto-created.
* We keep the core mechanic of determining the admin role solely by this flag, so we will ignore the role name.

Advantages:

* We don't need to override `::isAdmin()` and `::setIsAdmin()` or prevent their usage. They will work as designed.
* The intended behavior remains the same: when a new group type is created, we will have an admin role that has the role name "administrator" and it will have the flag set.
* Developers have the flexibility to create multiple admin roles, just like it can be done in core. I don't see many use cases for this, but there might be.
* Developers have the flexibility to revoke admin rights from the pre-created admin role, just like they can do in core, by calling `$role->setIsAdmin(FALSE)->save()`.

Disadvantages:

* It is possible to have a role that is called "administrator" but is in fact a regular role. This is a minor drawback, it matches the core functionality. It also requires a manual action from a developer to get into this state.

pfrenssen commented:
So I found this snippet in `OgAccess::userAccess()`:

```
      if ($membership = $this->membershipManager->getMembership($group, $user)) {
        foreach ($membership->getRoles() as $role) {
          // Check for the is_admin flag.
          /** @var \Drupal\og\Entity\OgRole $role */
          if ($role->isAdmin()) { // <<<< here it calls the inherited method
            $user_is_group_admin = TRUE;
            break;
          }

          $permissions = array_merge($permissions, $role->getPermissions());
        }
      }
```

It looks like this IS in fact implemented correctly. I also see that the flag is being set when the admin role is being created in `OgEventSubscriber::provideDefaultRoles()`:

```
    $role = $this->entityTypeManager->getStorage('og_role')->create([
      'name' => OgRoleInterface::ADMINISTRATOR,
      'label' => 'Administrator',
      'is_admin' => TRUE, // <<<<< Here it is being set.
    ]);
```

It also appears that I am the person who implemented this back in 2016 :confounded:

pfrenssen commented:
OK I found the source of my confusion for this PR. It appears that in our project we have exported our initial group types in May 2016, which is 1 month before this admin role functionality was implemented in OG. We simply have the wrong data exported.

Sorry for the noise. I will repurpose this PR to just update the test coverage for this.

pfrenssen commented:
I have written the additional test coverage that demonstrates that these methods actually work as designed. I've rebased the PR so this is just a single commit without the noise of the earlier implementation.

Test is running here: https://travis-ci.org/Gizra/og/builds/380706308

pfrenssen commented:
Test is green, up for review!

Ohmdesbois commented:
Bravo! and thank you !

pfrenssen commented:
You're welcome :)

amitaibu commented:
Looks good, thanks!
